ygot/diff_test.go

```diff
@@ -503,3 +487,553 @@ func TestFindSetLeaves(t *testing.T) {
+			}, {
+				Path: &gnmipb.Path{
+					Elem: []*gnmipb.PathElem{{
+						Name: "binary",
+					}},
+				},
+				Val: &gnmipb.TypedValue{&gnmipb.TypedValue_BytesVal{[]byte{42, 42, 42}}},
+			}, {
+				Path: &gnmipb.Path{
+					Elem: []*gnmipb.PathElem{{
+						Name: "empty",
+					}},
+				},
+				Val: &gnmipb.TypedValue{&gnmipb.TypedValue_BoolVal{true}},
+			}},
+		},
+	}, {
+		desc: "add an item to a list",
+		inOrig: &pathElemExample{
+			List: map[string]*pathElemExampleChild{
+				"p1": {Val: String("p1")},
+			},
+		},
+		inMod: &pathElemExample{
+			List: map[string]*pathElemExampleChild{
+				"p1": {Val: String("p1")},
+				"p2": {Val: String("p2")},
+			},
+		},
+		want: &gnmipb.Notification{
+			Update: []*gnmipb.Update{{
+				Path: &gnmipb.Path{
+					Elem: []*gnmipb.PathElem{{
+						Name: "list",
+						Key:  map[string]string{"val": "p2"},
+					}, {
+						Name: "val",
+					}},
+				},
+				Val: &gnmipb.TypedValue{Value: &gnmipb.TypedValue_StringVal{"p2"}},
+			}, {
+				Path: &gnmipb.Path{
+					Elem: []*gnmipb.PathElem{{
+						Name: "list",
+						Key:  map[string]string{"val": "p2"},
+					}, {
+						Name: "config",
+					}, {
+						Name: "val",
+					}},
+				},
+				Val: &gnmipb.TypedValue{Value: &gnmipb.TypedValue_StringVal{"p2"}},
+			}},
+		},
+	}, {
+		desc: "remove item from list",
+		inOrig: &pathElemExample{
+			List: map[string]*pathElemExampleChild{
+				"p1": {Val: String("p1")},
+				"p2": {Val: String("p2")},
+			},
+		},
+		inMod: &pathElemExample{
+			List: map[string]*pathElemExampleChild{
+				"p1": {Val: String("p1")},
+			},
+		},
+		want: &gnmipb.Notification{
+			Delete: []*gnmipb.Path{{
+				Elem: []*gnmipb.PathElem{{
+					Name: "list",
+					Key:  map[string]string{"val": "p2"},
+				}, {
+					Name: "val",
+				}},
+			}, {
+				Elem: []*gnmipb.PathElem{{
+					Name: "list",
+					Key:  map[string]string{"val": "p2"},
+				}, {
+					Name: "config",
+				}, {
+					Name: "val",
+				}},
+			}},
+		},
+	}, {
+		desc:    "invalid original",
+		inOrig:  &invalidGoStructEntity{},
+		inMod:   &invalidGoStructEntity{},
+		wantErr: String("could not extract set leaves from original struct"),
+	}, {
+		desc:   "invalid enum in modified",
+		inOrig: &badGoStruct{},
+		inMod: &badGoStruct{
+			InvalidEnum: 42,
+		},
+		wantErr: String("cannot represent field value 42 as TypedValue for path /an-enum"),
+	}, {
+		desc: "invalid enum in original",
+		inOrig: &badGoStruct{
+			InvalidEnum: 44,
+		},
+		inMod: &badGoStruct{
+			InvalidEnum: 42,
+		},
+		wantErr: String("cannot represent field value 42 as TypedValue for path /an-enum"),
+	}, {
+		desc:    "different types",
+		inOrig:  &renderExample{},
+		inMod:   &pathElemExample{},
+		wantErr: String("cannot diff structs of different types"),
+	}}
+
+	for _, tt := range tests {
+		got, err := Diff(tt.inOrig, tt.inMod)
+		if tt.wantErr != nil && err == nil || tt.wantErr != nil && !strings.Contains(err.Error(), *tt.wantErr) || tt.wantErr == nil && err != nil {
```

Comment:
(optional) Since we're doing a substring match now, consider calling the field wantErrSubStr or similar.

Reply:
Done.